Analysis does not continue after error (sonar.cxx.errorRecoveryEnabled=true)

Analysis does not continue after error (sonar.cxx.errorRecoveryEnabled=true)
https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Supported-configuration-properties#sonarcxxerrorrecoveryenabled

```
ERROR: Error during SonarQube Scanner execution
org.sonar.squidbridge.api.AnalysisException: Unable to analyze file: xxx.h
	at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:131)
	at org.sonar.cxx.sensors.squid.CxxSquidSensor.execute(CxxSquidSensor.java:164)
	at org.sonar.scanner.sensor.SensorWrapper.analyse(SensorWrapper.java:53)
        ...
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:63)
        ...
	at org.sonarsource.scanner.cli.Main.runAnalysis(Main.java:123)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:77)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.IllegalArgumentException: Line 41 is out of range for file xxx.h. File has 40 lines.
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at org.sonar.scanner.DefaultFileLinesContext.checkLineRange(DefaultFileLinesContext.java:70)
	at org.sonar.scanner.DefaultFileLinesContext.setIntValue(DefaultFileLinesContext.java:63)
	at org.sonar.cxx.sensors.visitors.CxxFileLinesVisitor.lambda$leaveFile$0(CxxFileLinesVisitor.java:185)
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.ReferencePipeline$Head.forEach(Unknown Source)
	at org.sonar.cxx.sensors.visitors.CxxFileLinesVisitor.leaveFile(CxxFileLinesVisitor.java:185)
	at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:71)
	at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:110)
	... 33 more
```

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/squid/CxxSquidSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/squid/CxxSquidSensor.java
@@ -130,7 +130,7 @@ public class CxxSquidSensor implements Sensor {
 
     List<SquidAstVisitor<Grammar>> visitors = new ArrayList<>((Collection) checks.all());
     visitors.add(new CxxHighlighterVisitor(context));
-    visitors.add(new CxxFileLinesVisitor(fileLinesContextFactory, context, linesOfCodeByFile));
+    visitors.add(new CxxFileLinesVisitor(language, fileLinesContextFactory, context, linesOfCodeByFile));
 
     visitors.add(
       new CxxCpdVisitor(

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/visitors/CxxFileLinesVisitor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/visitors/CxxFileLinesVisitor.java
@@ -39,9 +39,11 @@ import org.sonar.api.measures.FileLinesContext;
 import org.sonar.api.measures.FileLinesContextFactory;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
+import org.sonar.cxx.CxxLanguage;
 import org.sonar.cxx.api.CxxKeyword;
 import org.sonar.cxx.api.CxxPunctuator;
 import org.sonar.cxx.parser.CxxGrammarImpl;
+import org.sonar.cxx.sensors.utils.CxxUtils;
 import org.sonar.squidbridge.SquidAstVisitor;
 
 /**
@@ -51,6 +53,8 @@ import org.sonar.squidbridge.SquidAstVisitor;
 public class CxxFileLinesVisitor extends SquidAstVisitor<Grammar> implements AstAndTokenVisitor {
 
   private static final Logger LOG = Loggers.get(CxxFileLinesVisitor.class);
+
+  private final CxxLanguage language;
   private final FileLinesContextFactory fileLinesContextFactory;
   private static final Set<Integer> linesOfCode = Sets.newHashSet();
   private static final Set<Integer> linesOfComments = Sets.newHashSet();
@@ -74,8 +78,9 @@ public class CxxFileLinesVisitor extends SquidAstVisitor<Grammar> implements Ast
    * @param fileLinesContextFactory container for linesOfCode, linesOfComments, executableLines
    * @param allLinesOfCode set of lines for a source file
    */
-  public CxxFileLinesVisitor(FileLinesContextFactory fileLinesContextFactory, SensorContext context,
+  public CxxFileLinesVisitor(CxxLanguage language, FileLinesContextFactory fileLinesContextFactory, SensorContext context,
     Map<InputFile, Set<Integer>> allLinesOfCode) {
+    this.language = language;
     this.fileLinesContextFactory = fileLinesContextFactory;
     this.fileSystem = context.fileSystem();
     this.allLinesOfCode = allLinesOfCode;
@@ -182,11 +187,30 @@ public class CxxFileLinesVisitor extends SquidAstVisitor<Grammar> implements Ast
     }
     FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(inputFile);
 
-    linesOfCode.stream().forEach(line -> fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1));
-    linesOfComments.stream().forEach(line -> fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, 1));
-    executableLines.stream().forEach(line -> fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY,
-      line, 1));
-
+    try {
+      linesOfCode.stream().forEach(
+        line -> fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1)
+      );
+    } catch (IllegalArgumentException e) {
+      LOG.error("NCLOC_DATA_KEY metric error: {}", e.getMessage());
+      CxxUtils.validateRecovery(e, language);
+    }
+    try {
+      linesOfComments.stream().forEach(
+        line -> fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, 1)
+      );
+    } catch (IllegalArgumentException e) {
+      LOG.error("COMMENT_LINES_DATA_KEY metric error: {}", e.getMessage());
+      CxxUtils.validateRecovery(e, language);
+    }
+    try {
+      executableLines.stream().forEach(
+        line -> fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, 1)
+      );
+    } catch (IllegalArgumentException e) {
+      LOG.error("EXECUTABLE_LINES_DATA_KEY metric error: {}", e.getMessage());
+      CxxUtils.validateRecovery(e, language);
+    }
     fileLinesContext.save();
     this.allLinesOfCode.put(inputFile, Sets.newHashSet(linesOfCode));
 

--- a/cxx-sensors/src/test/java/org/sonar/cxx/sensors/visitors/CxxFileLinesVisitorTest.java
+++ b/cxx-sensors/src/test/java/org/sonar/cxx/sensors/visitors/CxxFileLinesVisitorTest.java
@@ -75,7 +75,7 @@ public class CxxFileLinesVisitorTest {
     when(fileLinesContextFactory.createFor(inputFile)).thenReturn(fileLinesContext);
 
     HashMap<InputFile, Set<Integer>> linesOfCode = new HashMap<>();
-    CxxFileLinesVisitor visitor = new CxxFileLinesVisitor(fileLinesContextFactory, sensorContext, linesOfCode);
+    CxxFileLinesVisitor visitor = new CxxFileLinesVisitor(language, fileLinesContextFactory, sensorContext, linesOfCode);
 
     CxxAstScanner.scanSingleFile(inputFile, sensorContext, TestUtils.mockCxxLanguage(), visitor);
 
@@ -109,7 +109,7 @@ public class CxxFileLinesVisitorTest {
     when(fileLinesContextFactory.createFor(inputFile)).thenReturn(fileLinesContext);
 
     HashMap<InputFile, Set<Integer>> linesOfCode = new HashMap<>();
-    CxxFileLinesVisitor visitor = new CxxFileLinesVisitor(fileLinesContextFactory, sensorContext, linesOfCode);
+    CxxFileLinesVisitor visitor = new CxxFileLinesVisitor(language, fileLinesContextFactory, sensorContext, linesOfCode);
 
     CxxAstScanner.scanSingleFile(inputFile, sensorContext, TestUtils.mockCxxLanguage(), visitor);
 

